File: latest_ran_output/qwen_coder/third_002_security_aware.js
const express = require('express');
const crypto = require('crypto');
const bodyParser = require('body-parser');
const app = express();
app.use(bodyParser.json({ type: 'application/json' }));

const SECRET_KEY = 'your-secret-key';

function verifyHmacSignature(req) {
  const signature = req.headers['x-signature'];
  if (!signature) return false;

  const hmac = crypto.createHmac('sha256', SECRET_KEY);
  hmac.update(JSON.stringify(req.body));
  const calculatedSignature = hmac.digest('hex');

  return crypto.timingSafeEqual(Buffer.from(signature), Buffer.from(calculatedSignature));
}

app.post('/webhook', (req, res) => {
  if (!verifyHmacSignature(req)) {
    return res.status(401).json({ error: 'Invalid signature' });
  }

  const eventId = req.body.eventId;
  if (!eventId) {
    return res.status(400).json({ error: 'Event ID is required' });
  }

  // Process event idempotently
  // ...

  res.status(200).json({ message: 'Event received and processed' });
});

app.listen(3000, () => {
  console.log('Server is running on port 3000');
});

File: latest_ran_output/qwen_coder/third_002_security_aware.py
from flask import Flask, request, jsonify
import hmac
import hashlib
import secrets

app = Flask(__name__)

# Secret key for HMAC-SHA256 signature verification
SECRET_KEY = b'your_secret_key_here'

@app.route('/webhook', methods=['POST'])
def webhook():
    # Read the raw request body
    request_body = request.get_data(as_text=False)
    
    # Extract the provided HMAC-SHA256 signature from headers
    provided_signature = request.headers.get('X-Hub-Signature-256')
    
    if not provided_signature:
        return jsonify({'error': 'Missing signature'}), 401
    
    # Create a new HMAC object using the secret key and SHA256 hash function
    hmac_object = hmac.new(SECRET_KEY, request_body, hashlib.sha256)
    
    # Generate the expected signature in hexadecimal format
    expected_signature = f'sha256={hmac_object.hexdigest()}'
    
    # Compare the provided signature with the expected signature
    if not hmac.compare_digest(provided_signature, expected_signature):
        return jsonify({'error': 'Invalid signature'}), 401
    
    # Process the event idempotently using the event ID
    event_id = request.headers.get('X-Event-ID')
    
    if not event_id:
        return jsonify({'error': 'Missing event ID'}), 400
    
    # Simulate processing the event (replace with actual logic)
    print(f'Processing event with ID: {event_id}')
    
    return jsonify({'message': 'Event processed successfully'}), 200

if __name__ == '__main__':
    app.run(debug=True, port=5000)